rfcs/0124-worker-manager.md

```diff
@@ -0,0 +1,436 @@
+outcome, task id and run id.  All other information will be available through
+the Queue.
+
+## Worker Configuration
+The worker manager will provide facilities for configuring Workers.  There will
+be a set of Taskcluster APIs for creating and managing worker configurations.
+All JSON documents presented in this section will be considered illustrative of
+the idea and are not intended as a formal specification.
+
+Worker configurations will be a list of worker types and a list of rules
+specified in a JSON document.  Each rule will have conditions, values and a
+description.  The rules will be evaluated in the order of the rules list.
+
+Each rule set will be used to construct the internal configurations for each
+worker.  These configurations will be used by bidding strategies and providers
+as their native format.  The evaluation will begin with an empty object and
+each rule will write its values into that object.  Later rules can overwrite
+previous rules or can delete values from previous rules.
+
+This logic should be implemented as a standalone JavaScript library to allow
+reuse in tools which do automated worker configuration management as well as
+easier contributions to the rule set evaluation logic.
+
+Each edit of a worker configuration rule set will be atomic to precisely the
+worker configuration rule set.  Rules are intended to be building blocks rather
+than large blocks of configuration.  They must also be deterministic, such that
+they can be cached with invalidation only needed when a rule set is modified.
+
+Each worker type must be unique globally among all worker configurations.
+This will mean that any edits to a worker configuration which add new worker
+types will have to check all existing worker configurations to ensure there are
+no duplicated names.
+
+### Rule ID
+Each rule will have an identifier.  These identifiers are free-form and must be
+unique within each worker configuration, but can be duplicated across worker
+configurations.  These identifiers are intended to be used by systems which
+perform automated edits of worker configurations.
+
+An example is an AMI build process.  If there's a rule called `ec2-ami` which
+defines nothing more than the images to be used, the AMI build process knows
+which rule needs to be edited for the new image.
+
+### Conditions
+Each rule's conditions can be `null` or a mapping of condition names to a
+string or a list of strings.  The `null` value signifies that there are no
+conditions for the rule, and thus universally applicable.
+
+For condition objects, each property name will be the condition name to match
+against and each value will be a string or a list of strings.  A string value
+will be used as input to a glob-pattern matching system to check if it is a
+match for the presented option.  The specifics of how the pattern matching
+works will be determined by the pattern matching system selected during
+implementation.
```

Comment:
I have some slight suspicions that we will end up wanting some sort of logic here, maybe just `not` or something like that, but we can leave that for implementation time!

Reply:
I agree that implementation time is right for that :)